src/FreeAge/server/game.cpp

```diff
@@ -1349,21 +1401,137 @@ bool Game::SimulateMeleeAttack(u32 /*unitId*/, ServerUnit* unit, u32 targetId, S
   return true;
 }
 
-void Game::ProduceUnit(ServerBuilding* building, UnitType unitInProduction) {
+void Game::ProduceUnit(u32 buildingId, ServerBuilding* building, UnitType unitInProduction) {
   // Create the unit object.
   u32 newUnitId;
   ServerUnit* newUnit = map->AddUnit(building->GetPlayerIndex(), unitInProduction, QPointF(-999, -999), &newUnitId);
   
+  QPointF freeSpace;
+  bool foundFreeSpace = FindFreeSpaceAroundBuilding(building, newUnit, freeSpace);
+  
+  if (foundFreeSpace) {
+    newUnit->SetMapCoord(freeSpace);
+  } else {
+    // TODO: set a special value that will be ignored by later checks 
+    newUnit->SetMapCoord(building->GetBaseTile());
+  }
+  
+  // Send messages to clients that see the new unit
+  QByteArray addObjectMsg = CreateAddObjectMessage(newUnitId, newUnit);
+  for (auto& player : *playersInGame) {
+    accumulatedMessages[player->index] += addObjectMsg;
+  }
+
+  if (!foundFreeSpace) {
+    // TODO: what happens when the capacity is full ?
+    building->GarrisonUnit(newUnit);
+    newUnit->SetGarrisonedInsideObject(building);
+
+    QByteArray unitGarrisonMessage = CreateUnitGarrisonMessage(newUnitId, buildingId);
+    for (auto& player : *playersInGame) {
+      accumulatedMessages[player->index] += unitGarrisonMessage;
+    }
+  }
+  
+  GetPlayerStats(newUnit->GetPlayerIndex())->UnitAdded(unitInProduction);
+}
+
+void Game::GarrisonUnit(u32 unitId, ServerUnit* unit, u32 targetObjectId, ServerObject* targetObject, bool enter, bool* unitMovementChanged) {
+  // NOTE: Could be simplified if Objects stored there ID. #ids
+  *unitMovementChanged = false;
+  if (enter) {
+    // TODO: test with target object being a unit
+
+    unit->StopMovement();
+    unit->RemoveTarget();
+
+    // Check if the object is a building and has <= 20% hp left
+    if (targetObject->GetGarrisonedUnitsCount() > 0 && targetObject->isBuilding() &&
```

Comment:
I think that `targetObject->GetGarrisonedUnitsCount() > 0` should be removed here.

Reply:
Why? Am I missing something?

Reply:
This piece of code is for the case that a unit wants to enter the building. In this case, it should not be allowed to enter the building if the building's HP is too low. But it should not matter whether there is already another unit inside the building or not.

There is another copy of this condition in the same file (for whether to ungarrison all units) where this check makes sense (since if the building is empty, there is nothing to ungarrison).

Reply:
yeah.... classic copy paste